File: torture-tests/language-coverage/java/CrossFileTaintTests.java
/**
 * =============================================================================
 * JAVA CROSS-FILE TAINT TRACKING TEST SUITE
 * =============================================================================
 *
 * PURPOSE: Test taint propagation across Java classes, packages, and layers.
 * Java's strong typing and OOP patterns create specific taint tracking
 * challenges across inheritance, interfaces, and dependency injection.
 *
 * STRUCTURE: Simulates a layered architecture with Controller -> Service ->
 * Repository pattern common in enterprise Java applications.
 *
 * =============================================================================
 */

package com.codescalpel.tests.crossfile;

import java.io.*;
import java.sql.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;
import java.util.stream.*;

// =============================================================================
// LAYER 1: DATA SOURCES (Simulates controller/input layer)
// =============================================================================

/**
 * UserInputSource - All methods return TAINTED data
 */
class UserInputSource {
    /**
     * TAINT SOURCE: HTTP query parameter
     */
    public String getQueryParam(String paramName) {
        return "tainted_query_" + paramName;
    }

    /**
     * TAINT SOURCE: Form data
     */
    public String getFormData(String fieldName) {
        return "tainted_form_" + fieldName;
    }

    /**
     * TAINT SOURCE: JSON body as Map
     */
    public Map<String, String> getJsonBody() {
        Map<String, String> body = new HashMap<>();
        body.put("userId", "tainted_user_id");
        body.put("action", "tainted_action");
        body.put("query", "tainted_query");
        return body; // All values TAINTED
    }

    /**
     * TAINT SOURCE: HTTP header
     */
    public String getHeader(String headerName) {
        return "tainted_header_" + headerName;
    }

    /**
     * TAINT SOURCE: Path variable
     */
    public String getPathVariable(String varName) {
        return "tainted_path_" + varName;
    }

    /**
     * TAINT SOURCE: User DTO from request
     */
    public UserDTO getUserDTO() {
        UserDTO dto = new UserDTO();
        dto.setId("tainted_id");
        dto.setName("tainted_name");
        dto.setEmail("tainted_email");
        dto.setRole("tainted_role");
        return dto; // All fields TAINTED
    }
}

/**
 * DTO class - fields can be tainted
 */
class UserDTO {
    private String id;
    private String name;
    private String email;
    private String role;

    public String getId() { return id; }
    public void setId(String id) { this.id = id; }
    public String getName() { return name; }
    public void setName(String name) { this.name = name; }
    public String getEmail() { return email; }
    public void setEmail(String email) { this.email = email; }
    public String getRole() { return role; }
    public void setRole(String role) { this.role = role; }
}

/**
 * ExternalDataSource - External data that should be untrusted
 */
class ExternalDataSource {
    /**
     * TAINT SOURCE: Message queue data
     */
    public String readFromQueue(String queueName) {
        return "tainted_queue_message";
    }

    /**
     * TAINT SOURCE: External API response
     */
    public Map<String, Object> callExternalApi(String url) {
        Map<String, Object> response = new HashMap<>();
        response.put("data", "tainted_api_data");
        return response;
    }

    /**
     * TAINT SOURCE: Database content (could contain stored XSS)
     */
    public String readFromDatabase(String query) {
        return "tainted_db_content";
    }
}

// =============================================================================
// LAYER 2: SERVICE LAYER (Business logic that preserves taint)
// =============================================================================

/**
 * DataProcessor - Transforms data but does NOT sanitize
 */
class DataProcessor {
    /**
     * TAINT PRESERVING: Validation does NOT remove taint
     */
    public String validateInput(String data) throws IllegalArgumentException {
        if (data == null || data.isEmpty()) {
            throw new IllegalArgumentException("Empty input");
        }
        if (data.length() > 10000) {
            throw new IllegalArgumentException("Input too long");
        }
        return data; // STILL TAINTED
    }

    /**
     * TAINT PRESERVING: Transformation does NOT sanitize
     */
    public String transformInput(String data) {
        return data.toUpperCase().trim(); // STILL TAINTED
    }

    /**
     * TAINT PRESERVING: Formatting does NOT sanitize
     */
    public String formatOutput(String data) {
        return String.format("[PROCESSED] %s", data); // STILL TAINTED
    }

    /**
     * TAINT PRESERVING: DTO field extraction
     */
    public String extractField(UserDTO dto, String fieldName) {
        switch (fieldName) {
            case "id": return dto.getId();
            case "name": return dto.getName();
            case "email": return dto.getEmail();
            case "role": return dto.getRole();
            default: return null;
        } // Returned value is TAINTED
    }

    /**
     * TAINT PRESERVING: Stream processing
     */
    public List<String> processStream(List<String> items) {
        return items.stream()
            .filter(s -> s != null && !s.isEmpty())
            .map(String::toLowerCase)
            .collect(Collectors.toList());
        // All items STILL TAINTED
    }
}

/**
 * DataTransformer - Multi-step pipeline
 */
class DataTransformer {
    private final DataProcessor processor;

    public DataTransformer(DataProcessor processor) {
        this.processor = processor;
    }

    /**
     * TAINT CHAIN: Full pipeline preserves taint
     */
    public String fullPipeline(String rawInput) {
        String validated = processor.validateInput(rawInput);
        String transformed = processor.transformInput(validated);
        String formatted = processor.formatOutput(transformed);
        return formatted; // STILL TAINTED after 3 transformations
    }

    /**
     * TAINT CHAIN: Async pipeline (simulated)
     */
    public CompletableFuture<String> asyncPipeline(String rawInput) {
        return CompletableFuture.supplyAsync(() -> processor.validateInput(rawInput))
            .thenApply(processor::transformInput)
            .thenApply(processor::formatOutput);
        // Result is STILL TAINTED
    }
}

/**
 * Generic service that preserves taint
 */
class GenericService<T, R> {
    private final Function<T, R> transformer;

    public GenericService(Function<T, R> transformer) {
        this.transformer = transformer;
    }

    /**
     * TAINT PRESERVING: Generic processing
     */
    public R process(T input) {
        return transformer.apply(input); // Taint flows through generic
    }
}

// =============================================================================
// LAYER 3: REPOSITORY/SINK LAYER
// =============================================================================

/**
 * DatabaseSink - SQL operations
 */
class DatabaseSink {
    /**
     * TAINT SINK: SQL Injection via WHERE clause
     */
    public void executeQuery(String whereClause) throws SQLException {
        String query = "SELECT * FROM users WHERE " + whereClause;
        System.out.println(query); // VULNERABILITY
    }

    /**
     * TAINT SINK: SQL Injection via INSERT
     */
    public void insertData(String table, String column, String value) throws SQLException {
        String query = String.format("INSERT INTO %s (%s) VALUES ('%s')", table, column, value);
        System.out.println(query); // VULNERABILITY
    }

    /**
     * TAINT SINK: SQL Injection via DTO fields
     */
    public void insertUser(UserDTO user) throws SQLException {
        String query = String.format(
            "INSERT INTO users (id, name, email, role) VALUES ('%s', '%s', '%s', '%s')",
            user.getId(), user.getName(), user.getEmail(), user.getRole()
        );
        System.out.println(query); // VULNERABILITY (4 injection points)
    }

    /**
     * TAINT SINK: SQL Injection via ORDER BY
     */
    public void queryWithSort(String sortColumn, String sortDirection) throws SQLException {
        String query = "SELECT * FROM users ORDER BY " + sortColumn + " " + sortDirection;
        System.out.println(query); // VULNERABILITY
    }
}

/**
 * CommandSink - Command execution
 */
class CommandSink {
    /**
     * TAINT SINK: Command Injection via Runtime.exec
     */
    public void executeCommand(String userArg) throws IOException {
        Runtime.getRuntime().exec("echo " + userArg); // VULNERABILITY
    }

    /**
     * TAINT SINK: Command Injection via ProcessBuilder
     */
    public void executeWithBuilder(String command) throws IOException {
        ProcessBuilder pb = new ProcessBuilder("sh", "-c", command);
        pb.start(); // VULNERABILITY
    }

    /**
     * TAINT SINK: Command Injection via array
     */
    public void executeWithArgs(String... args) throws IOException {
        Runtime.getRuntime().exec(args); // VULNERABILITY if args contains tainted data
    }
}

/**
 * FileSink - File operations
 */
class FileSink {
    /**
     * TAINT SINK: Path Traversal via File
     */
    public String readFile(String path) throws IOException {
        return new String(java.nio.file.Files.readAllBytes(
            java.nio.file.Paths.get(path)
        )); // VULNERABILITY
    }

    /**
     * TAINT SINK: Path Traversal via FileInputStream
     */
    public byte[] readWithStream(String path) throws IOException {
        try (FileInputStream fis = new FileInputStream(path)) { // VULNERABILITY
            return fis.readAllBytes();
        }
    }

    /**
     * TAINT SINK: Path Traversal via write
     */
    public void writeFile(String path, String content) throws IOException {
        java.nio.file.Files.writeString(
            java.nio.file.Paths.get(path), // VULNERABILITY
            content
        );
    }
}

/**
 * EvalSink - Code/expression evaluation
 */
class EvalSink {
    /**
     * TAINT SINK: Script Injection via ScriptEngine
     */
    public Object evaluateScript(String script) throws Exception {
        javax.script.ScriptEngine engine =
            new javax.script.ScriptEngineManager().getEngineByName("javascript");
        return engine.eval(script); // VULNERABILITY
    }

    /**
     * TAINT SINK: Expression Language Injection (simulated)
     */
    public void evaluateExpression(String expression) {
        // Simulates SpEL or OGNL evaluation
        System.out.println("Evaluating: " + expression); // VULNERABILITY
    }
}

// =============================================================================
// LAYER 4: INTEGRATION (Cross-layer taint flow)
// =============================================================================

/**
 * VulnerableApplication - Integrates all layers
 */
class VulnerableApplication {
    private final UserInputSource source;
    private final ExternalDataSource externalSource;
    private final DataProcessor processor;
    private final DataTransformer transformer;
    private final DatabaseSink dbSink;
    private final CommandSink cmdSink;
    private final FileSink fileSink;
    private final EvalSink evalSink;

    public VulnerableApplication() {
        this.source = new UserInputSource();
        this.externalSource = new ExternalDataSource();
        this.processor = new DataProcessor();
        this.transformer = new DataTransformer(processor);
        this.dbSink = new DatabaseSink();
        this.cmdSink = new CommandSink();
        this.fileSink = new FileSink();
        this.evalSink = new EvalSink();
    }

    /**
     * CROSS-FILE VULNERABILITY: Query param -> Service -> SQL
     */
    public void vulnerableSqlEndpoint(String paramName) throws SQLException {
        String userInput = source.getQueryParam(paramName);
        String validated = processor.validateInput(userInput);
        String transformed = processor.transformInput(validated);
        dbSink.executeQuery(transformed); // VULNERABILITY
    }

    /**
     * CROSS-FILE VULNERABILITY: Form data -> Format -> Command
     */
    public void vulnerableCommandEndpoint(String fieldName) throws IOException {
        String userInput = source.getFormData(fieldName);
        String formatted = processor.formatOutput(userInput);
        cmdSink.executeCommand(formatted); // VULNERABILITY
    }

    /**
     * CROSS-FILE VULNERABILITY: Path variable -> File read
     */
    public String vulnerableFileEndpoint(String varName) throws IOException {
        String userPath = source.getPathVariable(varName);
        return fileSink.readFile(userPath); // VULNERABILITY
    }

    /**
     * CROSS-FILE VULNERABILITY: JSON body -> Eval
     */
    public Object vulnerableEvalEndpoint() throws Exception {
        Map<String, String> body = source.getJsonBody();
        String expression = body.get("action");
        return evalSink.evaluateScript(expression); // VULNERABILITY
    }

    /**
     * CROSS-FILE VULNERABILITY: DTO through pipeline -> SQL
     */
    public void vulnerableDtoEndpoint() throws SQLException {
        UserDTO dto = source.getUserDTO();
        dbSink.insertUser(dto); // VULNERABILITY
    }

    /**
     * CROSS-FILE VULNERABILITY: Multi-hop pipeline
     */
    public void complexMultiHopVulnerability() throws SQLException {
        String raw = source.getQueryParam("search");
        String processed = transformer.fullPipeline(raw);
        dbSink.executeQuery(processed); // VULNERABILITY
    }

    /**
     * CROSS-FILE VULNERABILITY: Async pipeline -> SQL
     */
    public void asyncPipelineVulnerability() throws Exception {
        String raw = source.getQueryParam("async");
        String result = transformer.asyncPipeline(raw).get();
        dbSink.executeQuery(result); // VULNERABILITY
    }

    /**
     * CROSS-FILE VULNERABILITY: External API -> SQL
     */
    public void externalApiVulnerability() throws SQLException {
        Map<String, Object> response = externalSource.callExternalApi("http://api.example.com");
        String data = (String) response.get("data");
        dbSink.insertData("logs", "message", data); // VULNERABILITY
    }

    /**
     * CROSS-FILE VULNERABILITY: Queue message -> Command
     */
    public void queueToCommandVulnerability() throws IOException {
        String message = externalSource.readFromQueue("commands");
        cmdSink.executeCommand(message); // VULNERABILITY
    }

    /**
     * CROSS-FILE VULNERABILITY: Generic service chain
     */
    public void genericServiceVulnerability() throws IOException {
        String input = source.getQueryParam("generic");
        GenericService<String, String> service = new GenericService<>(String::toUpperCase);
        String result = service.process(input);
        cmdSink.executeCommand(result); // VULNERABILITY
    }

    /**
     * CROSS-FILE VULNERABILITY: Stream processing -> SQL
     */
    public void streamProcessingVulnerability() throws SQLException {
        List<String> inputs = Arrays.asList(
            source.getQueryParam("a"),
            source.getQueryParam("b"),
            source.getQueryParam("c")
        );
        List<String> processed = processor.processStream(inputs);
        for (String item : processed) {
            dbSink.insertData("logs", "value", item); // VULNERABILITY (3x)
        }
    }
}

// =============================================================================
// LAYER 5: INTERFACE-BASED TAINT PROPAGATION
// =============================================================================

/**
 * Interface for data sources
 */
interface DataSource {
    String getData(String key);
}

/**
 * Interface for data sinks
 */
interface DataSink {
    void consumeData(String data) throws Exception;
}

/**
 * Implementation with taint source
 */
class TaintedDataSource implements DataSource {
    @Override
    public String getData(String key) {
        return "tainted_" + key; // TAINT SOURCE
    }
}

/**
 * Implementation with taint sink
 */
class SqlDataSink implements DataSink {
    @Override
    public void consumeData(String data) {
        String query = "SELECT * FROM t WHERE x = '" + data + "'";
        System.out.println(query); // VULNERABILITY
    }
}

/**
 * Wiring through interfaces
 */
class InterfaceBasedApp {
    private final DataSource source;
    private final DataSink sink;

    public InterfaceBasedApp(DataSource source, DataSink sink) {
        this.source = source;
        this.sink = sink;
    }

    /**
     * CROSS-FILE VULNERABILITY: Interface-based taint flow
     */
    public void process(String key) throws Exception {
        String data = source.getData(key);
        sink.consumeData(data); // VULNERABILITY
    }
}

// =============================================================================
// LAYER 6: INHERITANCE-BASED TAINT PROPAGATION
// =============================================================================

/**
 * Abstract processor
 */
abstract class AbstractProcessor {
    /**
     * TAINT PRESERVING: Template method
     */
    public String process(String input) {
        String validated = validate(input);
        String transformed = transform(validated);
        return transformed; // TAINTED if input was tainted
    }

    protected abstract String validate(String input);
    protected abstract String transform(String input);
}

/**
 * Concrete processor that preserves taint
 */
class ConcreteProcessor extends AbstractProcessor {
    @Override
    protected String validate(String input) {
        return input.trim(); // STILL TAINTED
    }

    @Override
    protected String transform(String input) {
        return input.toUpperCase(); // STILL TAINTED
    }
}

/**
 * Inheritance-based app
 */
class InheritanceBasedApp {
    private final UserInputSource source;
    private final AbstractProcessor processor;
    private final DatabaseSink sink;

    public InheritanceBasedApp() {
        this.source = new UserInputSource();
        this.processor = new ConcreteProcessor();
        this.sink = new DatabaseSink();
    }

    /**
     * CROSS-FILE VULNERABILITY: Inheritance-based taint flow
     */
    public void process(String param) throws SQLException {
        String input = source.getQueryParam(param);
        String processed = processor.process(input);
        sink.executeQuery(processed); // VULNERABILITY
    }
}

// =============================================================================
// TEST RUNNER
// =============================================================================

public class CrossFileTaintTests {
    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("JAVA CROSS-FILE TAINT TRACKING TEST SUITE");
        System.out.println("=".repeat(60));
        System.out.println("");
        System.out.println("Layer Structure:");
        System.out.println("  1. Data Sources (UserInputSource, ExternalDataSource)");
        System.out.println("  2. Service Layer (DataProcessor, DataTransformer)");
        System.out.println("  3. Sink Layer (DatabaseSink, CommandSink, FileSink)");
        System.out.println("  4. Integration (VulnerableApplication)");
        System.out.println("  5. Interface-Based (DataSource, DataSink)");
        System.out.println("  6. Inheritance-Based (AbstractProcessor)");
        System.out.println("");
        System.out.println("Cross-File Taint Paths: 20+");
        System.out.println("Expected Vulnerabilities: 35");
        System.out.println("=".repeat(60));
    }
}

File: workflow-deep-security/language-coverage/typescript/cross_file_taint_tests.ts
/**
 * =============================================================================
 * TYPESCRIPT CROSS-FILE TAINT TRACKING TEST SUITE
 * =============================================================================
 *
 * PURPOSE: Test taint propagation across TypeScript files/modules with
 * specific focus on type system boundaries where taint tracking can fail.
 *
 * STRUCTURE: Simulates multiple modules with typed interfaces.
 * Type annotations do NOT affect runtime taint - must be tracked regardless.
 *
 * =============================================================================
 */

import { exec, execSync } from 'child_process';
import * as fs from 'fs';
import * as path from 'path';
import { EventEmitter } from 'events';

// =============================================================================
// TYPE DEFINITIONS
// =============================================================================

interface TaintedInput {
    value: string;
    source: 'query' | 'body' | 'header' | 'cookie' | 'path';
}

interface SafeInput {
    value: string;
    validated: true;
    sanitized: true;
}

interface UserData {
    id: string;
    name: string;
    email: string;
    role: string;
}

interface ApiResponse<T> {
    data: T;
    status: number;
    message: string;
}

type ProcessorFn<T, R> = (input: T) => R;
type AsyncProcessorFn<T, R> = (input: T) => Promise<R>;

// =============================================================================
// MODULE 1: TYPED DATA SOURCES
// =============================================================================

class TypedUserInputSource {
    /**
     * TAINT SOURCE: Returns typed but tainted data
     */
    getQueryParam(paramName: string): TaintedInput {
        return {
            value: `tainted_query_${paramName}`,
            source: 'query'
        };
    }

    /**
     * TAINT SOURCE: Request body with interface
     */
    getRequestBody<T extends object>(): T {
        // Type parameter doesn't affect taint
        return {
            userId: 'tainted_user_id',
            action: 'tainted_action'
        } as T; // TAINTED despite type assertion
    }

    /**
     * TAINT SOURCE: Typed header value
     */
    getHeader(headerName: string): string {
        return `tainted_header_${headerName}`;
    }

    /**
     * TAINT SOURCE: User data object
     */
    getUserData(): UserData {
        return {
            id: 'tainted_id',
            name: 'tainted_name',
            email: 'tainted_email',
            role: 'tainted_role'
        }; // All fields TAINTED
    }
}

class TypedExternalSource {
    /**
     * TAINT SOURCE: Generic API response
     */
    async fetchApi<T>(url: string): Promise<ApiResponse<T>> {
        return {
            data: { value: 'tainted_api_data' } as T,
            status: 200,
            message: 'ok'
        }; // data is TAINTED
    }

    /**
     * TAINT SOURCE: Database query result
     */
    async queryDatabase<T>(query: string): Promise<T[]> {
        return [{ content: 'tainted_db_content' }] as T[];
    }
}

// =============================================================================
// MODULE 2: TYPED DATA PROCESSORS
// =============================================================================

class TypedDataProcessor {
    /**
     * TAINT PRESERVING: Type conversion doesn't sanitize
     */
    convertToSafe(input: TaintedInput): SafeInput {
        // Type assertion does NOT sanitize!
        return {
            value: input.value, // STILL TAINTED
            validated: true,
            sanitized: true
        };
    }

    /**
     * TAINT PRESERVING: Generic processor
     */
    process<T, R>(input: T, transformer: ProcessorFn<T, R>): R {
        return transformer(input); // Taint flows through generic
    }

    /**
     * TAINT PRESERVING: Async generic processor
     */
    async processAsync<T, R>(
        input: T,
        transformer: AsyncProcessorFn<T, R>
    ): Promise<R> {
        return transformer(input); // Taint preserved through async
    }

    /**
     * TAINT PRESERVING: Partial type transformation
     */
    transformPartial<T extends object>(
        input: T,
        updates: Partial<T>
    ): T {
        return { ...input, ...updates }; // Taint from both sources
    }

    /**
     * TAINT PRESERVING: Pick specific fields
     */
    pickFields<T, K extends keyof T>(input: T, keys: K[]): Pick<T, K> {
        const result = {} as Pick<T, K>;
        for (const key of keys) {
            result[key] = input[key]; // Each field still tainted
        }
        return result;
    }
}

class TypedDataTransformer {
    private processor: TypedDataProcessor;

    constructor(processor: TypedDataProcessor) {
        this.processor = processor;
    }

    /**
     * TAINT CHAIN: Pipeline with types
     */
    fullPipeline(raw: TaintedInput): SafeInput {
        // This looks like it creates SafeInput but it's still tainted
        const safe = this.processor.convertToSafe(raw);
        return safe; // TAINTED despite SafeInput type
    }

    /**
     * TAINT CHAIN: Generic pipeline
     */
    genericPipeline<T, R>(input: T, ...transformers: ProcessorFn<any, any>[]): R {
        let current: any = input;
        for (const transformer of transformers) {
            current = transformer(current);
        }
        return current as R; // TAINTED through entire chain
    }
}

// =============================================================================
// MODULE 3: TYPED DATA SINKS
// =============================================================================

class TypedDatabaseSink {
    /**
     * TAINT SINK: SQL Injection with typed input
     */
    executeQuery(input: SafeInput): void {
        // Type says SafeInput but value is still tainted!
        const query = `SELECT * FROM users WHERE data = '${input.value}'`;
        console.log(query); // VULNERABILITY
    }

    /**
     * TAINT SINK: SQL with UserData fields
     */
    insertUser(user: UserData): void {
        const query = `INSERT INTO users (id, name, email, role) VALUES
            ('${user.id}', '${user.name}', '${user.email}', '${user.role}')`;
        console.log(query); // VULNERABILITY (4 injection points)
    }

    /**
     * TAINT SINK: Generic query execution
     */
    queryWithParams<T extends Record<string, string>>(
        table: string,
        params: T
    ): void {
        const conditions = Object.entries(params)
            .map(([k, v]) => `${k} = '${v}'`)
            .join(' AND ');
        const query = `SELECT * FROM ${table} WHERE ${conditions}`;
        console.log(query); // VULNERABILITY
    }
}

class TypedCommandSink {
    /**
     * TAINT SINK: Command Injection despite types
     */
    executeCommand(input: SafeInput): void {
        exec(input.value); // VULNERABILITY
    }

    /**
     * TAINT SINK: Typed array of commands
     */
    executeMultiple(inputs: SafeInput[]): void {
        for (const input of inputs) {
            execSync(input.value); // VULNERABILITY per element
        }
    }
}

class TypedFileSink {
    /**
     * TAINT SINK: Path Traversal with typed path
     */
    readFile(filepath: SafeInput): string {
        return fs.readFileSync(filepath.value, 'utf-8'); // VULNERABILITY
    }

    /**
     * TAINT SINK: Generic file operation
     */
    processFile<T extends { path: string }>(input: T): void {
        fs.readFileSync(input.path); // VULNERABILITY
    }
}

class TypedEvalSink {
    /**
     * TAINT SINK: Code Injection with type assertion
     */
    evaluate(input: SafeInput): unknown {
        return eval(input.value); // VULNERABILITY
    }

    /**
     * TAINT SINK: Function from typed string
     */
    createFunction(body: SafeInput): Function {
        return new Function(body.value); // VULNERABILITY
    }
}

// =============================================================================
// MODULE 4: TYPED INTEGRATION
// =============================================================================

class TypedVulnerableApplication {
    private source: TypedUserInputSource;
    private externalSource: TypedExternalSource;
    private processor: TypedDataProcessor;
    private transformer: TypedDataTransformer;
    private dbSink: TypedDatabaseSink;
    private cmdSink: TypedCommandSink;
    private fileSink: TypedFileSink;
    private evalSink: TypedEvalSink;

    constructor() {
        this.source = new TypedUserInputSource();
        this.externalSource = new TypedExternalSource();
        this.processor = new TypedDataProcessor();
        this.transformer = new TypedDataTransformer(this.processor);
        this.dbSink = new TypedDatabaseSink();
        this.cmdSink = new TypedCommandSink();
        this.fileSink = new TypedFileSink();
        this.evalSink = new TypedEvalSink();
    }

    /**
     * CROSS-FILE VULNERABILITY: Types don't prevent SQL Injection
     */
    vulnerableSqlWithTypes(param: string): void {
        const input: TaintedInput = this.source.getQueryParam(param);
        const safe: SafeInput = this.transformer.fullPipeline(input);
        this.dbSink.executeQuery(safe); // VULNERABILITY despite SafeInput type
    }

    /**
     * CROSS-FILE VULNERABILITY: UserData fields are tainted
     */
    vulnerableUserInsert(): void {
        const user: UserData = this.source.getUserData();
        this.dbSink.insertUser(user); // VULNERABILITY
    }

    /**
     * CROSS-FILE VULNERABILITY: Generic taint propagation
     */
    vulnerableGenericProcessing(): void {
        const input = this.source.getQueryParam('data');
        const processed = this.processor.process(
            input,
            (i: TaintedInput) => ({ value: i.value.toUpperCase() } as SafeInput)
        );
        this.cmdSink.executeCommand(processed); // VULNERABILITY
    }

    /**
     * CROSS-FILE VULNERABILITY: Async generic taint
     */
    async vulnerableAsyncGeneric(): Promise<void> {
        const input = this.source.getQueryParam('cmd');
        const result = await this.processor.processAsync(
            input,
            async (i: TaintedInput) => {
                await new Promise(r => setTimeout(r, 10));
                return { value: i.value, validated: true, sanitized: true } as SafeInput;
            }
        );
        this.evalSink.evaluate(result); // VULNERABILITY
    }

    /**
     * CROSS-FILE VULNERABILITY: API response to SQL
     */
    async vulnerableApiToSql(): Promise<void> {
        const response = await this.externalSource.fetchApi<{ query: string }>(
            'https://api.example.com'
        );
        const query = `SELECT * FROM t WHERE x = '${response.data}'`;
        console.log(query); // VULNERABILITY
    }

    /**
     * CROSS-FILE VULNERABILITY: Partial type doesn't sanitize
     */
    vulnerablePartialUpdate(): void {
        const user = this.source.getUserData();
        const updated = this.processor.transformPartial(user, { role: 'admin' });
        this.dbSink.insertUser(updated); // VULNERABILITY (id, name, email still tainted)
    }

    /**
     * CROSS-FILE VULNERABILITY: Pick preserves taint
     */
    vulnerablePickFields(): void {
        const user = this.source.getUserData();
        const picked = this.processor.pickFields(user, ['id', 'name']);
        const query = `SELECT * FROM t WHERE id = '${picked.id}' AND name = '${picked.name}'`;
        console.log(query); // VULNERABILITY
    }
}

// =============================================================================
// MODULE 5: TYPED EVENT-DRIVEN TAINT
// =============================================================================

interface TypedEvent<T> {
    type: string;
    payload: T;
    timestamp: number;
}

class TypedEventBus extends EventEmitter {
    /**
     * TAINT PRESERVING: Typed event emission
     */
    emitTyped<T>(event: TypedEvent<T>): void {
        this.emit(event.type, event.payload);
    }

    /**
     * TAINT SINK: Handler receives tainted payload
     */
    setupHandlers(): void {
        this.on('userInput', (payload: SafeInput) => {
            // Type says SafeInput but it's still tainted
            const query = `SELECT * FROM t WHERE x = '${payload.value}'`;
            console.log(query); // VULNERABILITY
        });

        this.on('command', (payload: { cmd: string }) => {
            exec(payload.cmd); // VULNERABILITY
        });
    }
}

class TypedEventApp {
    private source: TypedUserInputSource;
    private bus: TypedEventBus;

    constructor() {
        this.source = new TypedUserInputSource();
        this.bus = new TypedEventBus();
        this.bus.setupHandlers();
    }

    /**
     * CROSS-FILE VULNERABILITY: Typed event with tainted data
     */
    emitTaintedEvent(): void {
        const input = this.source.getQueryParam('data');
        const event: TypedEvent<SafeInput> = {
            type: 'userInput',
            payload: { value: input.value, validated: true, sanitized: true },
            timestamp: Date.now()
        };
        this.bus.emitTyped(event); // Taint flows to handler
    }
}

// =============================================================================
// MODULE 6: GENERIC CROSS-FILE TAINT
// =============================================================================

class GenericService<T, R> {
    private transformer: ProcessorFn<T, R>;

    constructor(transformer: ProcessorFn<T, R>) {
        this.transformer = transformer;
    }

    /**
     * TAINT PRESERVING: Generic processing
     */
    process(input: T): R {
        return this.transformer(input);
    }
}

class GenericRepository<T extends { id: string }> {
    /**
     * TAINT SINK: Generic query
     */
    findById(id: string): void {
        const query = `SELECT * FROM entities WHERE id = '${id}'`;
        console.log(query); // VULNERABILITY if id is tainted
    }

    /**
     * TAINT SINK: Generic insert
     */
    save(entity: T): void {
        const query = `INSERT INTO entities (id) VALUES ('${entity.id}')`;
        console.log(query); // VULNERABILITY if entity.id is tainted
    }
}

class GenericTaintApp {
    private source: TypedUserInputSource;

    constructor() {
        this.source = new TypedUserInputSource();
    }

    /**
     * CROSS-FILE VULNERABILITY: Generic service chain
     */
    processWithGenericService(): void {
        const input = this.source.getQueryParam('data');

        const service = new GenericService<TaintedInput, SafeInput>(
            (i) => ({ value: i.value, validated: true, sanitized: true })
        );

        const result = service.process(input);
        exec(result.value); // VULNERABILITY
    }

    /**
     * CROSS-FILE VULNERABILITY: Generic repository
     */
    useGenericRepository(): void {
        const user = this.source.getUserData();
        const repo = new GenericRepository<UserData>();
        repo.findById(user.id); // VULNERABILITY
        repo.save(user); // VULNERABILITY
    }
}

// =============================================================================
// TEST RUNNER
// =============================================================================

function runTypescriptCrossFileTaintTests(): void {
    console.log('='.repeat(60));
    console.log('TYPESCRIPT CROSS-FILE TAINT TRACKING TEST SUITE');
    console.log('='.repeat(60));
    console.log('');
    console.log('Module Structure:');
    console.log('  1. Typed Sources (TypedUserInputSource, TypedExternalSource)');
    console.log('  2. Typed Processors (TypedDataProcessor, TypedDataTransformer)');
    console.log('  3. Typed Sinks (TypedDatabaseSink, TypedCommandSink, etc.)');
    console.log('  4. Typed Integration (TypedVulnerableApplication)');
    console.log('  5. Typed Events (TypedEventBus, TypedEventApp)');
    console.log('  6. Generic Classes (GenericService, GenericRepository)');
    console.log('');
    console.log('Key Insight: TypeScript types provide NO runtime protection');
    console.log('Cross-File Taint Paths: 20+');
    console.log('Expected Vulnerabilities: 30');
    console.log('='.repeat(60));
}

export {
    // Types
    TaintedInput,
    SafeInput,
    UserData,
    // Sources
    TypedUserInputSource,
    TypedExternalSource,
    // Processors
    TypedDataProcessor,
    TypedDataTransformer,
    // Sinks
    TypedDatabaseSink,
    TypedCommandSink,
    TypedFileSink,
    TypedEvalSink,
    // Integration
    TypedVulnerableApplication,
    // Events
    TypedEventBus,
    TypedEventApp,
    // Generics
    GenericService,
    GenericRepository,
    GenericTaintApp,
    // Runner
    runTypescriptCrossFileTaintTests,
};
